### PR TITLE
sys/bhp_*: add initial support for generic Bottom Half Processor

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR addresses [this comment](https://github.com/RIOT-OS/RIOT/pull/18383#discussion_r940472395) and adds a generic Bottom Half Processor mechanism to be used by any device that requires ISR offloading.
The mechanism is agnostic to the underlying processing module (events, messages, thread flags), as it simply defines an interface for storing an (IRQ) handler and a context.

On top of that, a dedicated Bottom Half Processor module (`bhp_*`) exposes an init function that configures the underlying mechanism and setups the callback. Then, the device driver init function sets the ISR callback to `bhp_*_cb` and passes the `bhp_*_t` descriptor as the ISR context. The module will take care of the rest.

Based on this, I implemented an event based Bottom Half Processor (see `bhp_event`), which allows to offload the handler to an event queue. In the future, there could be more of these implementations (`bhp_thread_flags`, `bhp_msg`, `bhp_mbox`, etc).

### Testing procedure
Unittests should pass.

### Issues/PRs references
It will be come extremely handy to implement network-stack-and-device-agnostic IRQ offloaders for SPI radios (https://github.com/RIOT-OS/RIOT/pull/18383)

### CODE REVIEW

**Comment by jia200x** [comment]
I have a follow up PR that adapts #18383 to this mechanism. [See it in action](https://github.com/jia200x/RIOT/blob/pr/kw2xrf_bhp/sys/net/gnrc/netif/init_devs/auto_init_kw2xrf.c#L59)

File: sys/include/bhp/event.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * @param[in] cb        Bottom Half Processor callback
+ * @param[in] ctx       Context of @p cb
+ */
+void bhp_event_init(bhp_event_t *bhp_ev, event_queue_t *evq, bhp_cb_t cb, void *ctx);
+
+/**
+ * @brief Event based Bottom Half Processor ISR callback
+ *        To be called from ISR in order to trigger the Bottom Half Processor.
+ *
+ * @param[in] bhp_event_ctx     Context of the event based Bottom Half Processor.
```
Comment (benpicco) [inline sys/include/bhp/event.h:52]: Should be a pointer to `bhp_event_t` but we don't allow the implicit cast.

Comment (jia200x) [reply sys/include/bhp/event.h:52]: I made it that way on purpose to be able to hook it directly in `gpio_cb_t` callbacks or similar.

Comment (benpicco) [reply sys/include/bhp/event.h:52]: I know but it should be documented that we expect the argument to be a `bhp_event_t *`

Comment (jia200x) [reply sys/include/bhp/event.h]: ah, true. Done!

File: sys/bhp/Kconfig [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+menu "Bottom Half Processor"
+
+config MODULE_BHP
+    bool
+
+config MODULE_BHP_EVENT
+    bool "Enable event based Bottom Half Processor implementation"
+    depends on MODULE_EVENT
+    select MODULE_BHP
```
Comment (benpicco) [inline sys/bhp/Kconfig:16]: Seems a bit redundant to define both?
@leandrolanzieri might know better

Comment (leandrolanzieri) [reply sys/bhp/Kconfig:16]: I think it is relatively low-risk here to `select MODULE_EVENT`, if that's what you mean

Comment (jia200x) [reply sys/bhp/Kconfig:16]: Hmmm now that I see it, probably will need to change the modelling a bit.
My idea was to use `MODULE_BHP` as a feature request that the driver can select.
Then, another module (e.g `gnrc_netif`) can select `MODULE_BHP_EVENT` **if** `MODULE_BHP` is present. That way we don't pull BHP code when devices don't require it

Comment (jia200x) [reply sys/bhp/Kconfig]: I just had an offline discussion with @leandrolanzieri and we agreed on leaving `MODULE_BHP` to compile `sys/bhp`. This is also used to indicate in Make that a driver requires a Bottom Half Processor, **but** it doesn't force the user to bring any of the `bhp_*` modules.
For Kconfig I simply added a `HAVE_BHP_IRQ_HANDLER` feature to the driver that indicates there's an IRQ handler expected to be offloaded.

File: drivers/kw2xrf/Kconfig [hunk 3]
```diff
@@ -21,6 +21,7 @@ menuconfig MODULE_KW2XRF
     select MODULE_NETDEV
     select MODULE_NETDEV_IEEE802154
     select MODULE_CORE_THREAD_FLAGS
+    select HAVE_BHP_IRQ_HANDLER
```
Comment (benpicco) [inline drivers/kw2xrf/Kconfig:24]: Not yet, right? 😉

Comment (jia200x) [reply drivers/kw2xrf/Kconfig:24]: Well, there is a "BHP Handler" expected to be called by someone (either any of the `bhp_*` modules or a custom based solution).
Maybe I can find another name to make it less confusing

Comment (jia200x) [reply drivers/kw2xrf/Kconfig:24]: ah! sorry, wrong PR :D

Comment (jia200x) [reply drivers/kw2xrf/Kconfig]: done

**Comment by jia200x** [comment]
if this is merged before #18383, I can adapt the kinetis radio to this BHP mechanism

**Review by benpicco — APPROVED** [review]
Let’s do this then

**Comment by jia200x** [comment]
thanks for the review!

**Comment by kaspar030** [comment]
> The mechanism is agnostic to the underlying processing module (events, messages, thread flags), as it simply defines an interface for storing an (IRQ) handler and a context.

Is it that simple? The memory model of the four methods is very different.

**Comment by jia200x** [comment]
> > The mechanism is agnostic to the underlying processing module (events, messages, thread flags), as it simply defines an interface for storing an (IRQ) handler and a context.
> 
> Is it that simple? The memory model of the four methods is very different.

It is agnostic from an interface point of view. Neither the device nor the network stack care about which underlying mechanism is being used and these can be exchanged at any time.

**Pull request merged**